src/Microsoft.AspNetCore.JsonPatch/Internal/ParsedPath.cs

```diff
@@ -12,14 +14,19 @@ public struct ParsedPath
 
         private readonly string[] _segments;
 
+        private static readonly Regex escapingRegex = new Regex("~(?<code>0|1)", RegexOptions.Compiled);
+
         public ParsedPath(string path)
         {
             if (path == null)
             {
                 throw new ArgumentNullException(nameof(path));
             }
 
-            _segments = path.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
+            _segments = path
+                .Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries)
+                .Select(segment => escapingRegex.Replace(segment, (match) => match.Groups["code"].Value == "0" ? "~" : "/"))
```

Comment:
Can we change this to not use regex? It's a pretty big hammer (and super slow and inefficient), so it might be nicer to either use `string.Split`, or do a more manual conversion. I think the code is as simple as `<path>.Replace("~1", "/").Replace("~0", "~")`. Note that the order of these operations is significant because you don't want to double-decode the `~` character.